File: src/delay_node.cpp
#include <ros/ros.h>
#include <geometry_msgs/TwistStamped.h>
#include <message_filters/time_sequencer.h>
#include <message_filters/subscriber.h>

using namespace geometry_msgs;

ros::Publisher pub;

void delay_cb(const geometry_msgs::TwistStampedConstPtr &delay_vel)
{
    ROS_INFO("Delay Callback");
    TwistStamped new_vel;
    new_vel.header.stamp = ros::Time::now();
    new_vel.twist.angular.z = delay_vel->twist.angular.z;
    new_vel.twist.linear.x = delay_vel->twist.linear.x;
    pub.publish(new_vel);
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "delay_node");
    ros::NodeHandle n;
    // Params
    double delay_time;
    std::string delay_topic;
    std::string new_topic;

    n.param<std::string>("/jackal_bsc/delay_topic", delay_topic, "/jackal_bsc/key_vel_stamped");
    n.param<std::string>("/jackal_bsc/new_topic", new_topic, "/jackal_bsc/key_vel_delay");
    n.param("/jackal_bsc/delay_time", delay_time, 1.0);

    pub = n.advertise<TwistStamped>(new_topic, 10);

    // Time Sequencer
    message_filters::Subscriber<TwistStamped> sub(n, delay_topic, 1);
    message_filters::TimeSequencer<geometry_msgs::TwistStamped> seq(sub, ros::Duration(delay_time),
                                                     ros::Duration(0.01), 10);
    seq.registerCallback(delay_cb);

    ROS_INFO("Delay node created sequencer and is spinning...");
    ros::spin();
    return 0;
}

File: scripts/delay_node.py
#!/usr/bin/env python

import functools
import rospy
from geometry_msgs.msg import TwistStamped, Twist


class DelayNode:
    def __init__(self):
        self.delay_time = rospy.get_param("/jackal_bsc/delay_time", 1.0)
        delay_topic = rospy.get_param("/jackal_bsc/delay_topic", "/jackal_bsc/key_vel_stamped")
        new_topic = rospy.get_param("/jackal_bsc/new_topic", "/jackal_bsc/key_vel_delay")

        self.delay_pub = rospy.Publisher(new_topic, TwistStamped, queue_size=10)
        sub = rospy.Subscriber(delay_topic, TwistStamped, self.callback, queue_size=40)

        # For testing
        # delay_topic = '/jackal_bsc/key_vel'
        # new_topic = '/jackal_bsc/key_vel_delay'
        # self.delay_pub = rospy.Publisher(new_topic, Twist, queue_size=10)
        # sub = rospy.Subscriber(delay_topic, Twist, self.callback, queue_size=40)

    def delayed_callback(self, msg, event):
        # Change header time to sync with nav_vel
        msg.header.stamp = rospy.Time().now()
        self.delay_pub.publish(msg)

    def callback(self, msg):
        timer = rospy.Timer(rospy.Duration(self.delay_time),
                            functools.partial(self.delayed_callback, msg),
                            oneshot=True)

    def start(self):
        rospy.loginfo("[Delay]: Delay node started")
        rospy.spin()


if __name__ == '__main__':
    rospy.init_node("delay_node")
    d = DelayNode()
    d.start()
